ee/tabby-webserver/src/schema/email.rs

```diff
@@ -1,11 +1,70 @@
-use anyhow::Result;
+use anyhow::{anyhow, Result};
 use async_trait::async_trait;
-use juniper::GraphQLObject;
+use juniper::{GraphQLEnum, GraphQLObject};
+use tabby_common::{enum_conversion::DatabaseSerializable, enum_mapping};
+use validator::validate_url;
+
+#[derive(GraphQLEnum, Clone, Debug)]
+pub enum Encryption {
+    StartTls,
+    SslTls,
+    None,
+}
+
+enum_mapping!(
+    Encryption:
+    StartTls => "starttls",
+    SslTls => "ssltls",
+    None => "none",
+);
+
+#[derive(GraphQLEnum, Clone, Debug)]
+pub enum AuthMethod {
+    Plain,
+    Login,
+    XOAuth2,
+}
+
+enum_mapping!(
+    AuthMethod:
+    Plain => "plain",
+    Login => "login",
+    XOAuth2 => "xoauth2",
+);
 
 #[derive(GraphQLObject)]
 pub struct EmailSetting {
     pub smtp_username: String,
     pub smtp_server: String,
+    pub from_address: String,
+    pub encryption: Encryption,
+    pub auth_method: AuthMethod,
+}
+
+impl EmailSetting {
+    pub fn new_validate(
+        smtp_username: String,
+        smtp_server: String,
+        from_address: Option<String>,
+        encryption: String,
+        auth_method: String,
+    ) -> Result<Self> {
+        if !validate_url(&smtp_server) {
+            return Err(anyhow!("Invalid smtp server address"));
+        }
+
+        let encryption = Encryption::from_db_str(&encryption)?;
+        let auth_method = AuthMethod::from_db_str(&auth_method)?;
+
+        let from_address = from_address.unwrap_or_else(|| smtp_username.clone());
```

Comment:
when from_address is null, return error.

Reply:
Why not just take `String` instead of `Option<String>` then?

Reply:
that works as well. but since new_validate takes data from db side, it does have a chance being null (and you'll need to check it before calling try_from then)

Reply:
Right. I had thought we wanted to allow this to be None too, and maybe just emit a warning about it rather than an error. An error would cause the entire settings object to be discarded and no email could be sent, so if we make it an error the migration immediately breaks your settings.

Reply:
discussed in slack